src/checkpoint.rs

```diff
@@ -0,0 +1,141 @@
+use crate::errors::Result;
+use crate::mock::MockStorage;
+use crate::traits::{Api, Extern, ReadonlyStorage, Storage};
+
+pub struct Checkpoint<'a, S: Storage> {
+    /// a backing storage that is only modified upon commit
+    storage: &'a mut S,
+    /// these are local changes not flushed to backing storage
+    local_state: MockStorage,
+    /// this is a list of changes to be written to backing storage upon commit
+    rep_log: Vec<Op>,
+}
+
+enum Op {
+    Set { key: Vec<u8>, value: Vec<u8> },
+}
+
+impl<'a, S: Storage> Checkpoint<'a, S> {
```

Comment:
With this `commit` and `rollback` functionality, this is analogue to a typical database transaction. I think we should rename `Checkpoint` -> `Transaction`. Maybe better use `StorageTransaction` to avoid confusion with blockchain transactions.

Reply:
It is a sub transaction.

The closest example I have see is SAVEPOINT in postgresql
https://www.postgresql.org/docs/9.1/sql-savepoint.html
https://www.postgresql.org/docs/9.1/sql-rollback-to.htm

Transaction is overloaded and I try to not reuse it.
SubTransaction maybe?

Other ideas?

Reply:
"Sub transaction" or "transaction inside transaction" seem to be implementation details. What I think is important is the need to explicitely commit in order to persist the changes.

`StorageTransaction` makes most sense to me to differentiate between a blokchain transaction and a storage transaction. `SubTransaction` is irritating when you don't have any main transaction.

Reply:
Okay, then StorageTransaction for the struct.
And `transactional()` for the wrapper function?